### ISSUE TITLE
The binary operator Equal is not defined for the types 'System.Nullable`1[System.Boolean]' and 'System.Boolean' error for queries with complex predicates involving nullable boolean properties

### ISSUE THREAD

**Issue by mo-esmp** [aspnet/EntityFramework#7061]
I've updated EF from ver 1.0.1 to ver 1.1.0 and after runnig below query I got error.
```
var data = await _context.CustomerServices
                .Include(a => a.ServiceEquipment)
                .Include(a => a.ServiceAddresses)
                .ThenInclude(a => a.Address)
                .SingleOrDefaultAsync(a => a.Id == serviceId && !a.IsDeleted && !a.ServiceEquipment.IsDeleted);
```

> the binary operator Equal is not defined for the types 'System.Nullable`1[System.Boolean]' and 'System.Boolean'.
> 
>    at System.Linq.Expressions.Expression.GetEqualityComparisonOperator(ExpressionType binaryType, String opName, Expression left, Expression right, Boolean liftToNull)
>    at System.Linq.Expressions.Expression.Equal(Expression left, Expression right, Boolean liftToNull, MethodInfo method)
>    at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.SearchConditionTranslatingVisitor.VisitUnary(UnaryExpression expression)
>    at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.SearchConditionTranslatingVisitor.VisitBinary(BinaryExpression expression)
>    at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.SearchConditionTranslatingVisitor.VisitUnary(UnaryExpression expression)
>    at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.ApplyOptimizations(Expression expression, Boolean searchCondition, Boolean joinCondition)
>    at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.VisitSelect(SelectExpression selectExpression)
>    at Remotion.Linq.Parsing.ThrowingExpressionVisitor.Visit(Expression expression)
>    at Microsoft.EntityFrameworkCore.Query.Sql.DefaultQuerySqlGenerator.GenerateSql(IReadOnlyDictionary`2 parameterValues)
>    at Microsoft.EntityFrameworkCore.Query.Internal.ShaperCommandContext.GetRelationalCommand(IReadOnlyDictionary`2 parameters)
>    at Microsoft.EntityFrameworkCore.Query.Internal.AsyncQueryingEnumerable.AsyncEnumerator.<BufferlessMoveNext>d__9.MoveNext()
> --- End of stack trace from previous location where exception was thrown ---
>    at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    at Microsoft.EntityFrameworkCore.Storage.Internal.SqlServerExecutionStrategy.<ExecuteAsync>d__6`2.MoveNext()

when I remove `!a.ServiceEquipment.IsDeleted` from query it works.

**Comment by maumar** [comment]
@mo-esmp can you share the model you are using? Specifically code listing for all the entities and contents of OnModelCreating method on the DbContext

**Comment by mo-esmp** [comment]
Entites:
```
    public class BaseEntity : ISoftDeleteTrackable
    {
        public BaseEntity()
        {
        }

        public int Id { get; set; }

        /// <summary>
        /// Gets or sets a value indicating whether this instance is deleted. 
        /// </summary>
        /// <value> <c> true </c> if this instance is deleted; otherwise, <c> false </c>. </value>
        public bool IsDeleted { get; set; }
    }

   public class CustomerServiceEntity : BaseEntity
    {
        public CustomerServiceStatus Status { get; set; }

        public BillingCycle ServiceDuration { get; set; }

        public decimal Price { get; set; }

        public int? CreatorId { get; set; }

        public DateTime CreatedDate { get; set; }

        public int CustomerId { get; set; }

        public CustomerEntity Customer { get; set; }

        public CustomerOwnServiceEquipmentEntity ServiceEquipment { get; set; }

        public ICollection<CustomerServiceAddressEntity> ServiceAddresses { get; set; }
    }

    public class CustomerOwnServiceEquipmentEntity : BaseEntity
    {
        public int NoOfFloors { get; set; }

        public bool HasTower { get; set; }

        public float? TowerHeight { get; set; }

        public bool HasEquipment { get; set; }

        public string EquipmentType { get; set; }

        public CustomerServiceEntity Service { get; set; }
    }

    public class CustomerServiceAddressEntity : BaseEntity
    {
        public CustomerServiceAddressType Type { get; set; }

        public int AddressId { get; set; }

        public AddressEntity Address { get; set; }

        public int ServiceId { get; set; }

        public CustomerServiceEntity Service { get; set; }
    }

    public class AddressEntity : BaseEntity
    {
        public string Phone { get; set; }

        public string Address { get; set; }

        public int CityId { get; set; }

        public string CityName { get; set; }

        public int ProvinceId { get; set; }

        public string ProvinceName { get; set; }

        public string PostalCode { get; set; }

        public string AreaCode { get; set; }

        public ICollection<CustomerServiceAddressEntity> ServiceAddresses { get; set; }
    }
```

Maps:
```
public class CustomerServiceMap : IEntityMap
{
    public CustomerServiceMap(ModelBuilder builder)
    {
        builder.Entity<CustomerServiceEntity>(e =>
        {
            e.Property(g => g.Status).IsRequired();
            e.Property(g => g.ServiceDuration).IsRequired();
            e.Property(g => g.Price).HasColumnType("decimal(18, 2)").IsRequired();
            e.Property(g => g.StartDate).IsRequired(false);
            e.Ignore(g => g.ServiceDurationTitle);
            e.Ignore(g => g.StatusTitle);

            e.HasOne(s => s.ServiceEquipment).WithOne(cp => cp.Service)
                .HasPrincipalKey<CustomerServiceEntity>(cs => cs.Id)
                .HasForeignKey<CustomerOwnServiceEquipmentEntity>(cose => cose.Id);

            e.HasMany(s => s.ServiceAddresses).WithOne(cp => cp.Service).HasForeignKey(s => s.ServiceId);
        });
    }
}

public class CustomerOwnServiceEquipmentMap : IEntityMap
{
    public CustomerOwnServiceEquipmentMap(ModelBuilder builder)
    {
        builder.Entity<CustomerOwnServiceEquipmentEntity>(e =>
        {
            e.Property(m => m.HasEquipment).IsRequired();
            e.Property(m => m.HasTower).IsRequired();
            e.Property(m => m.NoOfFloors).IsRequired();
            e.Property(m => m.TowerHeight).IsRequired(false);
            e.Property(m => m.EquipmentType).HasColumnType("nvarchar(512)").IsRequired(false);
        });
    }
}

public class CustomerServiceAddressMap : IEntityMap
{
    public CustomerServiceAddressMap(ModelBuilder builder)
    {
        builder.Entity<CustomerServiceAddressEntity>(e =>
        {
            e.Property(csa => csa.Type).IsRequired();
            e.Property(csa => csa.AddressId).IsRequired();
            e.Property(csa => csa.ServiceId).IsRequired();
        });
    }
}

public class AddressMap : IEntityMap
{
    public AddressMap(ModelBuilder builder)
    {
        builder.Entity<AddressEntity>(e =>
        {
            e.Property(m => m.CityId).IsRequired();
            e.Property(m => m.ProvinceId).HasDefaultValue(1).IsRequired();
            e.Property(m => m.Phone).HasColumnType("varchar(11)").IsRequired(false);
            e.Property(m => m.CityName).HasColumnType("nvarchar(50)").IsRequired(false);
            e.Property(m => m.ProvinceName).HasColumnType("nvarchar(50)").IsRequired(false);
            e.Property(m => m.PostalCode).HasColumnType("nvarchar(10)").IsRequired(false);
            e.Property(m => m.AreaCode).HasColumnType("char(3)").IsRequired(false);
            e.Property(m => m.Address).HasColumnType("nvarchar(512)").IsRequired();
            e.Property(m => m.RepresentativeId).IsRequired(false);
        });
    }
}
```
```
protected override void OnModelCreating(ModelBuilder builder)
{
    RegisterMaps(builder);

    base.OnModelCreating(builder);
}

private void RegisterMaps(ModelBuilder builder)
{
    var maps = Assembly.GetExecutingAssembly()
        .GetTypes()
        .Where( type => !string.IsNullOrWhiteSpace(type.Namespace) && typeof(IEntityMap).IsAssignableFrom(type) && type.IsClass)
        .ToList();

    foreach (var item in maps)
        Activator.CreateInstance(item, BindingFlags.Public | BindingFlags.Instance, null, new object[] { builder }, null);
}
```

**Comment by maumar** [comment]
Simplified repro:
```cs
    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureCreated();
                ctx.CustomerServices.SingleOrDefault(a => a.Id == 1 && !a.ServiceEquipment.IsDeleted);
            }
        }
    }


    public class MyContext : DbContext
    {
        public DbSet<CustomerServiceEntity> CustomerServices { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer("Data Source=.;Initial Catalog=Repro7061;Integrated Security=SSPI;");
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<CustomerServiceEntity>(e =>
            {
                e.HasOne(s => s.ServiceEquipment).WithOne(cp => cp.Service)
                    .HasPrincipalKey<CustomerServiceEntity>(cs => cs.Id)
                    .HasForeignKey<CustomerOwnServiceEquipmentEntity>(cose => cose.Id);
            });
        }
    }

    public class CustomerServiceEntity
    {
        public int Id { get; set; }

        public CustomerOwnServiceEquipmentEntity ServiceEquipment { get; set; }
    }

    public class CustomerOwnServiceEquipmentEntity
    {
        public int Id { get; set; }
        public bool IsDeleted { get; set; }

        public CustomerServiceEntity Service { get; set; }
    }
```

**Comment by maumar** [comment]
Fixed in a7e0117695f23fe49c9d0a8b7f69209f5f3ec394

**Comment by divega** [comment]
@maumar was this fixed pushed into the 1.1.1? Should we reopen the issue for approval?

**Comment by Eilon** [comment]
This patch bug is **approved**. Please use the normal code review process w/ a PR and make sure the fix is in the correct branch, then close the bug and mark it as done.

**Comment by bjcull** [comment]
FYI, for anyone else waiting for this to be released, you can get around it with a good ol' fashioned hack:

    a.ServiceEquipment.IsDeleted == (bool?)false

**Comment by anurse** [comment]
Verified that this is fixed in the 1.1.1 candidate build.
